R/validation.R

```diff
@@ -0,0 +1,207 @@
+TARGET_VARIABLE <- 'yyyyy'
+FORMULA <- as.formula(paste(TARGET_VARIABLE, "~ ."))
+K <- 8 # Because I have 4 cores :/
+
+#### Training/Testing ######
+
+get_assignments <- function(survey_df, k) {
+  sample(rep(1:k, length.out=nrow(survey_df)))
+}
+
+
+split_test_train <- function(survey_df, k=K) {
+  assignments <- get_assignments(survey_df, k)
+  folds <- lapply(1:k, function(i) {
+    is_test <- assignments == i
+    test <- survey_df[is_test, ]
+    train <- survey_df[!is_test, ]
+    list(test=test, train=train, fold_number=i)
+  })
+  folds
+}
+
+
+test_one <- function(method, fold) {
+  stopifnot(length(method) == 1)
+  method_name <- names(method)
+  method <- method[[1]]
+  predictions <- method(fold)
+  df <- data.frame(
+    true=fold$test[, TARGET_VARIABLE],
+    predicted=predictions,
+    fold=fold$fold_number)
+  if (!is.null(method_name)) {
+    df$method <- method_name
+  }
+  df
+}
+
+
+test_method_on_splits <- function(method, folds) {
+  method_name <- names(method)
+  parallel <- (method_name != 'tpo') # RWeka does not work with multicore
+  if (parallel) {
+    mclapply(
+      folds,
+      function(fold) test_one(method, fold),
+      mc.cores=detectCores())
+  } else {
+    lapply(
+      folds,
+      function(fold) test_one(method, fold))
+  }
+}
+
+
+test_all <- function(survey_df, method_list=METHOD_LIST, k=K) {
+  folds <- split_test_train(survey_df, k)
+  method_results <- purrr::flatmap(names(method_list), function(method_name) {
+    print(method_name)
+    method <- method_list[method_name]
+    predictions <- test_method_on_splits(method, folds)
+  })
+  purrr::reduce(method_results, rbind)
+}
+
+##### Output #####
+DEFAULT_THRESHOLDS <- seq(0.1, 0.4, .1)
+
+add_threshold <- function(output, threshold) {
+  consumption_cutoff <- quantile(output$true, threshold)
+  output <- output[rep(1:nrow(output), each=length(threshold)), ]
+  output$threshold <- as.integer(threshold * 100)
+  output$consumption_cutoff <- consumption_cutoff
+  output
+}
+
+
+order_by_predicted <- function(output) {
+  output %>% 
+    arrange(predicted) %>%
+    mutate(pct_targeted=row_number() / n())
+}
+
+
+reach_by_pct_targeted <- function(output, threshold=DEFAULT_THRESHOLDS) {
+  output %>%
+    add_threshold(threshold) %>%
+    group_by(method, threshold) %>%
+    order_by_predicted() %>%
+    mutate(tp=true < consumption_cutoff) %>%
+    mutate(value=cumsum(tp) / n())
+}
+
+
+value_at_pct <- function(stat_by_pct) {
+  filter(stat_by_pct, pct_targeted <= threshold / 100) %>%
```

Comment:
I'm going to leave it as it is since if this function does not return results ordered by pct_targeted it's a good indication something isn't working. I will add some tests to verify these functions before we run though since you're right it's tricky to get these right.